clientconn.go

```diff
@@ -830,90 +831,98 @@ func (ac *addrConn) waitForStateChange(ctx context.Context, sourceState Connecti
-		}
-		newTransport, err := transport.NewClientTransport(ctx, sinfo, ac.dopts.copts)
-		// Don't call cancel in success path due to a race in Go 1.6:
-		// https://github.com/golang/go/issues/15078.
-		if err != nil {
-			cancel()
-
-			if e, ok := err.(transport.ConnectionError); ok && !e.Temporary() {
-				return err
+		for _, addr := range ac.addrs {
+			ac.mu.Lock()
+			ac.printf("connecting")
+			if ac.state == Shutdown {
+				// ac.tearDown(...) has been invoked.
+				ac.mu.Unlock()
+				return errConnClosing
+			}
+			if ac.down != nil {
+				ac.down(downErrorf(false, true, "%v", errNetworkIO))
+				ac.down = nil
+			}
+			ac.state = Connecting
+			ac.stateCV.Broadcast()
+			t := ac.transport
+			ac.mu.Unlock()
+			if closeTransport && t != nil {
+				t.Close()
+			}
+			timeout := minConnectTimeout
+			if timeout < sleepTime {
+				timeout = sleepTime
+			}
+			ctx, cancel := context.WithTimeout(ac.ctx, timeout)
+			sinfo := transport.TargetInfo{
+				Addr:     addr.Addr,
+				Metadata: addr.Metadata,
+			}
+			newTransport, err := transport.NewClientTransport(ctx, sinfo, ac.dopts.copts)
+			// Don't call cancel in success path due to a race in Go 1.6:
+			// https://github.com/golang/go/issues/15078.
+			if err != nil {
+				cancel()
+
+				if e, ok := err.(transport.ConnectionError); ok && !e.Temporary() {
+					return err
+				}
+				grpclog.Warningf("grpc: addrConn.resetTransport failed to create client transport: %v; Reconnecting to %v", err, addr)
+				ac.mu.Lock()
+				if ac.state == Shutdown {
+					// ac.tearDown(...) has been invoked.
+					ac.mu.Unlock()
+					return errConnClosing
+				}
+				ac.errorf("transient failure: %v", err)
+				ac.state = TransientFailure
+				ac.stateCV.Broadcast()
+				if ac.ready != nil {
+					close(ac.ready)
+					ac.ready = nil
+				}
+				ac.mu.Unlock()
+				closeTransport = false
+				select {
+				case <-ac.ctx.Done():
+					return ac.ctx.Err()
+				default:
+				}
+				continue
 			}
-			grpclog.Warningf("grpc: addrConn.resetTransport failed to create client transport: %v; Reconnecting to %v", err, ac.addr)
 			ac.mu.Lock()
+			ac.printf("ready")
 			if ac.state == Shutdown {
 				// ac.tearDown(...) has been invoked.
 				ac.mu.Unlock()
+				newTransport.Close()
 				return errConnClosing
 			}
-			ac.errorf("transient failure: %v", err)
-			ac.state = TransientFailure
+			ac.state = Ready
 			ac.stateCV.Broadcast()
+			ac.transport = newTransport
 			if ac.ready != nil {
 				close(ac.ready)
 				ac.ready = nil
 			}
-			ac.mu.Unlock()
-			closeTransport = false
-			timer := time.NewTimer(sleepTime - time.Since(connectTime))
-			select {
-			case <-timer.C:
-			case <-ac.ctx.Done():
-				timer.Stop()
-				return ac.ctx.Err()
+			if ac.cc.dopts.balancer != nil {
+				ac.down = ac.cc.dopts.balancer.Up(addr)
 			}
-			timer.Stop()
-			continue
-		}
-		ac.mu.Lock()
-		ac.printf("ready")
-		if ac.state == Shutdown {
-			// ac.tearDown(...) has been invoked.
+			ac.curAddr = addr
```

Comment:
Initiate it each time as "0" at the beginning of resetTransport seems better. It also avoids the race when other thread want to use conn[current address] to find current addrconn.